openhtf/io/frontend/__main__.py

```diff
@@ -32,97 +32,207 @@
+gflags.DEFINE_integer('discovery_port',
+                      multicast.DEFAULT_PORT,
+                      'Multicast port to use for station discovery.')
+gflags.DEFINE_integer('discovery_ttl',
+                      multicast.DEFAULT_TTL,
+                      'TTL for station discovery pings.')
+
+
+_LOG = logging.getLogger(__name__)
+
+UNKNOWN_STATION_ID = 'UNKNOWN_STATION'
+
+
+class Station(mutablerecords.Record(  # pylint: disable=too-few-public-methods,no-init
+    'Station', [], {'station_id': UNKNOWN_STATION_ID, 'state': None})):
+  """Represents a station seen on the local network.
+
+  Attributes:
+    station_id: Station's ID string.
+    state: A dictionary representation of station state with two keys:
+      framework: A dictionary representation of a TestExecutor object.
+      test: A dictionary representation of a TestState object.
+    """
+
+
+class StationStore(threads.KillableThread):
+  """Self-updating store of stations visible on the local network(s).
+
+  Station data is stored in the 'stations' attribute, a dictionary mapping
+  tuples of (host_ip, port) to Station records.
+  """
+  def __init__(self):
+    super(StationStore, self).__init__()
+    self._live = False
+    self.hostname = socket.gethostname()
+    self.stations = {}
+
+  @staticmethod
+  def MessageFramework(host, port, method='GET', message=None):
+    """Interact with a running instance of openhtf via its HTTP API.
+
+    Args:
+      host: The IP address of the host running openhtf.
+      port: The TCP port of openhtf's HTTP API on host.
+      method: 'GET' or 'POST'.
+      message: Prompt response for POSTs. Ignored for GETs.
+    """
+    result = None
+    conn = httplib.HTTPConnection(host, int(port))
+    args = [method, '/']
+    if message and method == 'POST':
+      args.append(message)
+    try:
+      conn.request(*args)
+      if method == 'GET':
+        response = conn.getresponse()
+        if response.status != 200:
+          result = None
+        else:
+          result = json.loads(response.read())
+    except socket.error as e:
+      result = None
+    conn.close()
+    return result
+
+  def _Discover(self):
+    """Use multicast to discover stations on the local network."""
+    responses = multicast.send(multicast.PING_STRING,
+                               FLAGS.discovery_address,
+                               FLAGS.discovery_port,
+                               FLAGS.discovery_ttl)
+    for host, response in responses:
+      try:
+        port = json.loads(response)[multicast.PING_RESPONSE_KEY]
+        self._Track(host, port)
+      except (KeyError, ValueError):
+        _LOG.warning('Ignoring unrecognized discovery response from %s: %s' % (
+            host, response))
+
+  def _ThreadProc(self):
+    """Continuously scan for new stations and add them to the store."""
+    self._live = True
+    while self._live:
+      self._Discover()
+      time.sleep(FLAGS.discovery_interval_s)
```

Comment:
if you want to get fancy, you can use a threading event and wait() with a timeout, and if it gets set that's your 'thread killed' signal - ie instead of self._live = False in Stop(), you do self._killed_event.set().  this way, if you have a long discovery interval, your join() in Stop() can take a while.